File: scripts/debug_rope.py
#!/usr/bin/env python3
"""
Print step-by-step values of RoPE computation in diffusers
to compare with Rust implementation.
"""
import torch
import math
import numpy as np


def prepare_video_coords_debug(
    batch_size: int,
    num_frames: int,
    height: int,
    width: int,
    rope_interpolation_scale,
    base_num_frames: int = 20,
    base_height: int = 2048,
    base_width: int = 2048,
    patch_size: int = 1,
    patch_size_t: int = 1,
):
    """Exact recreation of diffusers LTXVideoRotaryPosEmbed._prepare_video_coords"""
    device = "cpu"
    
    grid_h = torch.arange(height, dtype=torch.float32, device=device)
    grid_w = torch.arange(width, dtype=torch.float32, device=device)
    grid_f = torch.arange(num_frames, dtype=torch.float32, device=device)
    
    print(f"grid_f: {grid_f[:5]}...")
    print(f"grid_h: {grid_h[:5]}...")
    print(f"grid_w: {grid_w[:5]}...")
    
    grid = torch.meshgrid(grid_f, grid_h, grid_w, indexing="ij")
    grid = torch.stack(grid, dim=0)  # (3, T, H, W)
    grid = grid.unsqueeze(0).repeat(batch_size, 1, 1, 1, 1)  # (B, 3, T, H, W)
    
    print(f"grid shape after meshgrid: {grid.shape}")
    print(f"grid[0, 0, :, 0, 0] (T coords at H=0, W=0): {grid[0, 0, :, 0, 0]}")
    print(f"grid[0, 1, 0, :, 0] (H coords at T=0, W=0): {grid[0, 1, 0, :, 0][:5]}...")
    print(f"grid[0, 2, 0, 0, :] (W coords at T=0, H=0): {grid[0, 2, 0, 0, :][:5]}...")
    
    if rope_interpolation_scale is not None:
        grid[:, 0:1] = grid[:, 0:1] * rope_interpolation_scale[0] * patch_size_t / base_num_frames
        grid[:, 1:2] = grid[:, 1:2] * rope_interpolation_scale[1] * patch_size / base_height
        grid[:, 2:3] = grid[:, 2:3] * rope_interpolation_scale[2] * patch_size / base_width
    
    print(f"\nAfter scaling:")
    print(f"grid[0, 0, :, 0, 0] (T coords scaled): {grid[0, 0, :, 0, 0]}")
    print(f"grid[0, 1, 0, :3, 0] (H coords scaled): {grid[0, 1, 0, :3, 0]}")
    print(f"grid[0, 2, 0, 0, :3] (W coords scaled): {grid[0, 2, 0, 0, :3]}")
    
    grid = grid.flatten(2, 4).transpose(1, 2)  # (B, T*H*W, 3)
    
    print(f"\nAfter flatten and transpose: {grid.shape}")
    print(f"First 5 positions (t,h,w coords):")
    for i in range(min(5, grid.shape[1])):
        print(f"  pos {i}: {grid[0, i, :]}")
    
    return grid


def compute_freqs_debug(grid, dim=2048, theta=10000.0):
    """Exact recreation of diffusers LTXVideoRotaryPosEmbed.forward frequency computation"""
    
    freq_dim = dim // 6
    print(f"\nFrequency computation:")
    print(f"dim={dim}, freq_dim={freq_dim}, theta={theta}")
    
    start = 1.0
    end = theta
    # freqs = theta^linspace(log_theta(start), log_theta(end), freq_dim)
    freqs = theta ** torch.linspace(
        math.log(start, theta),
        math.log(end, theta),
        freq_dim,
        device=grid.device,
        dtype=torch.float32,
    )
    print(f"freqs (before *pi/2): {freqs[:5]}...")
    
    freqs = freqs * math.pi / 2.0
    print(f"freqs (after *pi/2): {freqs[:5]}...")
    
    # freqs * (grid * 2 - 1)
    # grid: (B, L, 3), freqs: (freq_dim,)
    # grid.unsqueeze(-1): (B, L, 3, 1)
    # grid * 2 - 1: scale to [-1, 1]
    grid_scaled = grid.unsqueeze(-1) * 2 - 1
    print(f"\ngrid_scaled shape: {grid_scaled.shape}")
    print(f"grid_scaled first pos: {grid_scaled[0, 0, :, 0]}")
    
    freqs_full = freqs * grid_scaled  # (B, L, 3, freq_dim)
    print(f"freqs_full shape: {freqs_full.shape}")
    
    freqs_full = freqs_full.transpose(-1, -2).flatten(2)  # (B, L, 3*freq_dim)
    print(f"freqs after transpose&flatten: {freqs_full.shape}")
    print(f"First 10 freqs at pos 0: {freqs_full[0, 0, :10]}")
    
    cos_freqs = freqs_full.cos().repeat_interleave(2, dim=-1)
    sin_freqs = freqs_full.sin().repeat_interleave(2, dim=-1)
    
    print(f"\ncos_freqs shape: {cos_freqs.shape}")
    print(f"First 10 cos values at pos 0: {cos_freqs[0, 0, :10]}")
    print(f"sin_freqs shape: {sin_freqs.shape}")
    print(f"First 10 sin values at pos 0: {sin_freqs[0, 0, :10]}")
    
    # Padding if dim % 6 != 0
    if dim % 6 != 0:
        pad_size = dim % 6
        cos_padding = torch.ones_like(cos_freqs[:, :, :pad_size])
        sin_padding = torch.zeros_like(sin_freqs[:, :, :pad_size])
        cos_freqs = torch.cat([cos_padding, cos_freqs], dim=-1)
        sin_freqs = torch.cat([sin_padding, sin_freqs], dim=-1)
        print(f"\nAfter padding (dim%6={pad_size}):")
        print(f"cos_freqs final shape: {cos_freqs.shape}")
    
    return cos_freqs, sin_freqs


def main():
    # Parameters matching our test run
    batch_size = 1
    num_frames = 4  # latent frames (25 video frames / 8)
    height = 10     # latent height (320 / 32)
    width = 16      # latent width (512 / 32)
    dim = 2048      # hidden_size
    
    # rope_interpolation_scale from diffusers pipeline
    frame_rate = 25.0
    vae_temporal_compression = 8
    vae_spatial_compression = 32
    rope_interpolation_scale = (
        vae_temporal_compression / frame_rate,  # 8/25 = 0.32
        vae_spatial_compression,                 # 32
        vae_spatial_compression,                 # 32
    )
    
    print("=" * 60)
    print("Diffusers RoPE Debug")
    print("=" * 60)
    print(f"num_frames={num_frames}, height={height}, width={width}")
    print(f"rope_interpolation_scale={rope_interpolation_scale}")
    print()
    
    grid = prepare_video_coords_debug(
        batch_size, num_frames, height, width, rope_interpolation_scale
    )
    
    cos_freqs, sin_freqs = compute_freqs_debug(grid, dim=dim)
    
    print("\n" + "=" * 60)
    print("Summary")
    print("=" * 60)
    print(f"Final cos_freqs range: [{cos_freqs.min():.4f}, {cos_freqs.max():.4f}]")
    print(f"Final sin_freqs range: [{sin_freqs.min():.4f}, {sin_freqs.max():.4f}]")
    
    # Save for Rust comparison
    np.save("output/diffusers_cos_freqs.npy", cos_freqs.numpy())
    np.save("output/diffusers_sin_freqs.npy", sin_freqs.numpy())
    print(f"\nSaved freqs to output/diffusers_cos_freqs.npy and output/diffusers_sin_freqs.npy")


if __name__ == "__main__":
    main()

File: src/bin/debug_rope.rs
//! Debug binary for comparing Rust RoPE output with diffusers
//! Run with: cargo run --release --bin debug-rope

use candle_core::{DType, Device};
use candle_video::rope::{FractionalRoPE, generate_indices_grid_for_diffusers};
use std::fs::File;
use std::io::Write;

fn main() -> anyhow::Result<()> {
    println!("{}", "=".repeat(60));
    println!("Rust RoPE Debug");
    println!("{}", "=".repeat(60));

    let device = Device::Cpu; // Use CPU for comparison
    let dtype = DType::F32;

    // Parameters matching our test run
    let batch_size = 1;
    let num_frames = 4; // latent frames
    let height = 10; // latent height
    let width = 16; // latent width
    let dim = 2048; // hidden_size (head_dim for RoPE)
    let seq_len = num_frames * height * width;

    // RoPE parameters from diffusers
    let frame_rate = 25.0_f32;
    let vae_temporal_compression = 8.0_f32;
    let vae_spatial_compression = 32.0_f32;

    let rope_scale_t = vae_temporal_compression / frame_rate; // 8/25 = 0.32
    let rope_scale_h = vae_spatial_compression; // 32
    let rope_scale_w = vae_spatial_compression; // 32

    println!(
        "num_frames={}, height={}, width={}",
        num_frames, height, width
    );
    println!(
        "rope_interpolation_scale=({:.2}, {:.0}, {:.0})",
        rope_scale_t, rope_scale_h, rope_scale_w
    );
    println!();

    // Generate indices grid
    let indices_grid = generate_indices_grid_for_diffusers(
        batch_size,
        num_frames,
        height,
        width,
        20,   // base_num_frames
        2048, // base_height
        2048, // base_width
        1,    // patch_size
        1,    // patch_size_t
        rope_scale_t,
        rope_scale_h,
        rope_scale_w,
        &device,
    )?;

    println!("indices_grid shape: {:?}", indices_grid.dims());

    // Print first few positions (indices_grid is (B, 3, L))
    let grid_f32 = indices_grid.to_dtype(DType::F32)?;
    let grid_vec: Vec<f32> = grid_f32.flatten_all()?.to_vec1()?;

    println!("\nFirst 5 positions (t,h,w coords at dim 1):");
    for i in 0..5.min(seq_len) {
        let t_idx = i;
        let h_idx = seq_len + i;
        let w_idx = 2 * seq_len + i;
        println!(
            "  pos {}: [{:.4}, {:.4}, {:.4}]",
            i, grid_vec[t_idx], grid_vec[h_idx], grid_vec[w_idx]
        );
    }

    // Compute RoPE frequencies
    let rope = FractionalRoPE::new(dim, 10000.0, seq_len);
    let (cos_freqs, sin_freqs) = rope.compute_freqs_cis(&indices_grid.to_dtype(dtype)?, &device)?;

    println!("\ncos_freqs shape: {:?}", cos_freqs.dims());
    println!("sin_freqs shape: {:?}", sin_freqs.dims());

    // Get first 10 values at position 0
    let cos_vec: Vec<f32> = cos_freqs.to_dtype(DType::F32)?.flatten_all()?.to_vec1()?;
    let sin_vec: Vec<f32> = sin_freqs.to_dtype(DType::F32)?.flatten_all()?.to_vec1()?;

    println!(
        "\nFirst 10 cos values at pos 0: {:?}",
        &cos_vec[0..10.min(cos_vec.len())]
    );
    println!(
        "First 10 sin values at pos 0: {:?}",
        &sin_vec[0..10.min(sin_vec.len())]
    );

    // Compute ranges
    let cos_min = cos_vec.iter().cloned().fold(f32::INFINITY, f32::min);
    let cos_max = cos_vec.iter().cloned().fold(f32::NEG_INFINITY, f32::max);
    let sin_min = sin_vec.iter().cloned().fold(f32::INFINITY, f32::min);
    let sin_max = sin_vec.iter().cloned().fold(f32::NEG_INFINITY, f32::max);

    println!("\n{}", "=".repeat(60));
    println!("Summary");
    println!("{}", "=".repeat(60));
    println!("Final cos_freqs range: [{:.4}, {:.4}]", cos_min, cos_max);
    println!("Final sin_freqs range: [{:.4}, {:.4}]", sin_min, sin_max);

    // Save for comparison
    println!("\nSaving to output/rust_rope_*.bin for comparison...");
    std::fs::create_dir_all("output")?;

    let mut f = File::create("output/rust_cos_freqs.bin")?;
    for v in &cos_vec {
        f.write_all(&v.to_le_bytes())?;
    }

    let mut f = File::create("output/rust_sin_freqs.bin")?;
    for v in &sin_vec {
        f.write_all(&v.to_le_bytes())?;
    }

    println!("Done! Compare with Python:");
    println!("  python scripts/compare_rope.py");

    Ok(())
}
